File: algorithms/sorting/cycle-sort/scala/CycleSort.scala
package algorithms.sorting.cycle

/**
 * Cycle Sort implementation.
 * An in-place, unstable sorting algorithm that is optimal in terms of
 * the number of writes to the original array.
 */
object CycleSort {
  def sort(arr: Array[Int]): Array[Int] = {
    val result = arr.clone()
    val n = result.length

    for (cycleStart <- 0 until n - 1) {
      var item = result(cycleStart)

      var pos = cycleStart
      for (i <- cycleStart + 1 until n) {
        if (result(i) < item) {
          pos += 1
        }
      }

      if (pos != cycleStart) {
        while (item == result(pos)) {
          pos += 1
        }

        if (pos != cycleStart) {
          val temp = item
          item = result(pos)
          result(pos) = temp
        }

        while (pos != cycleStart) {
          pos = cycleStart
          for (i <- cycleStart + 1 until n) {
            if (result(i) < item) {
              pos += 1
            }
          }

          while (item == result(pos)) {
            pos += 1
          }

          if (item != result(pos)) {
            val temp = item
            item = result(pos)
            result(pos) = temp
          }
        }
      }
    }

    result
  }
}

File: algorithms/sorting/cycle-sort/csharp/CycleSort.cs
using System;

namespace Algorithms.Sorting.Cycle
{
    /**
     * Cycle Sort implementation.
     * An in-place, unstable sorting algorithm that is optimal in terms of
     * the number of writes to the original array.
     */
    public static class CycleSort
    {
        public static int[] Sort(int[] arr)
        {
            if (arr == null)
            {
                return new int[0];
            }

            int[] result = (int[])arr.Clone();
            int n = result.Length;

            for (int cycleStart = 0; cycleStart <= n - 2; cycleStart++)
            {
                int item = result[cycleStart];

                int pos = cycleStart;
                for (int i = cycleStart + 1; i < n; i++)
                {
                    if (result[i] < item)
                    {
                        pos++;
                    }
                }

                if (pos == cycleStart)
                {
                    continue;
                }

                while (item == result[pos])
                {
                    pos++;
                }

                if (pos != cycleStart)
                {
                    int temp = item;
                    item = result[pos];
                    result[pos] = temp;
                }

                while (pos != cycleStart)
                {
                    pos = cycleStart;
                    for (int i = cycleStart + 1; i < n; i++)
                    {
                        if (result[i] < item)
                        {
                            pos++;
                        }
                    }

                    while (item == result[pos])
                    {
                        pos++;
                    }

                    if (item != result[pos])
                    {
                        int temp = item;
                        item = result[pos];
                        result[pos] = temp;
                    }
                }
            }

            return result;
        }
    }
}
